Sources/Flatten.swift

```diff
@@ -733,6 +754,124 @@ private struct LatestState<Value, Error: Swift.Error> {
+			let composite = CompositeDisposable()
+			let relayDisposable = CompositeDisposable()
+
+			composite += relayDisposable
+			composite += self.observeFirst(observer, relayDisposable)
+
+			return composite
+		}
+	}
+
+	fileprivate func observeFirst(_ observer: Observer<Value.Value, Error>, _ relayDisposable: CompositeDisposable) -> Disposable? {
+		let state = Atomic(FirstState<Value.Value, Error>())
+
+		return self.observe { event in
+			switch event {
+			case let .value(innerProducer):
+				// Ignore consecutive `innerProducer`s if first `innerSignal` already sent an event.
+				guard !relayDisposable.isDisposed else {
+					return
+				}
+
+				innerProducer.startWithSignal { innerSignal, innerDisposable in
+
+					let disposableHandle = relayDisposable.add(innerDisposable)
+
+					innerSignal.observe { [unowned innerSignal] event in
+
+						let isFirstSignal: Bool = state.modify {
+							if $0.active == nil {
+								$0.active = innerSignal
+								$0.innerSignalComplete = false
+							}
+							return $0.active === innerSignal
+						}
+
+						// Ignore non-winning signals.
+						guard isFirstSignal else { return }
+
+						// Dispose all running innerSignals except winning one.
+						disposableHandle.remove()
+						relayDisposable.dispose()
+
+						switch event {
+						case .completed:
+							let shouldComplete: Bool = state.modify {
+								$0.innerSignalComplete = true
+								return $0.outerSignalComplete
+							}
+
+							if shouldComplete {
+								observer.sendCompleted()
+							}
+
+						case .value, .failed, .interrupted:
+							observer.action(event)
+						}
+					}
+				}
+
+			case let .failed(error):
+				observer.send(error: error)
+
+			case .completed:
+				let shouldComplete: Bool = state.modify {
+					$0.outerSignalComplete = true
+					return $0.innerSignalComplete
+				}
+
+				if shouldComplete {
+					observer.sendCompleted()
+				}
+
+			case .interrupted:
+				observer.sendInterrupted()
+			}
+		}
+	}
+}
+
+extension SignalProducerProtocol where Value: SignalProducerProtocol, Error == Value.Error {
+	/// Returns a signal that forwards values from the first signal sent on
+	/// `self`, ignoring values sent on consecutive inner signals.
+	///
+	/// An error sent on `self` or the first inner signal will be sent on the
+	/// returned signal.
+	///
+	/// The returned signal completes when `self` and the first inner
+	/// signal have both completed.
+	fileprivate func first() -> SignalProducer<Value.Value, Error> {
+		return SignalProducer<Value.Value, Error> { observer, disposable in
+			let relayDisposable = CompositeDisposable()
+			disposable += relayDisposable
+
+			self.startWithSignal { signal, signalDisposable in
+				disposable += signalDisposable
+				disposable += signal.observeFirst(observer, relayDisposable)
+			}
+		}
+	}
+}
+
+private struct FirstState<Value, Error: Swift.Error> {
+	var outerSignalComplete: Bool = false
+	var innerSignalComplete: Bool = true
```

Comment:
It seems a little weird to default this to `true`. It might be better to default to `false` and check for `innerSignalComplete || active == nil` in `shouldComplete`?